examples/async/s3_cp.ml

```diff
@@ -0,0 +1,390 @@
+    (* This corresponds to p.21 of the s3 api doc
+       we're making:
+       <HTTPMethod>\n
+       <CanonicalURI>\n
+       <CanonicalQueryString>\n
+       <CanonicalHeaders>\n
+       <SignedHeaders>\n
+       <HashedPayload>
+    *)
+    let open Cohttp.Request in
+    let http_method = Code.string_of_method request.meth in
+    (* Nb the path will be url encoded as per spec *)
+    let canoncical_uri = Compat.encode_string(Uri.path request.uri) in
+    (* Sort query string in alphabetical order by key *)
+    let canonical_query = Compat.encode_query_string request.uri in
+    let sorted_headers = Header.to_list request.headers
+                         |> List.sort ~cmp:ksrt in
+    let canonical_headers = sorted_headers
+                            |> List.fold ~init:"" ~f:(fun acc (k,v) ->
+                                acc ^
+                                (Printf.sprintf "%s:%s\n"
+                                   (String.lowercase k) (String.strip v)))
+    in
+    let signed_headers = sorted_headers
+                         |> List.map ~f:(fun (k,_) -> k)
+                         |> String.concat ~sep:";"
+    in
+    (Printf.sprintf "%s\n%s\n%s\n%s\n%s\n%s"
+       http_method canoncical_uri canonical_query canonical_headers signed_headers
+       hashed_payload, signed_headers)
+
+  let string_to_sign ?time ~scope ~service canonical_request:string =
+    (* As per p. 23 of s3 api doc. The requests need current time in utc
+       time parameter is there for testing. *)
+    let time_str = match time with
+        None -> Time.to_string_abs ~zone:Time.Zone.utc (Time.now())
+      | Some t -> Compat.format_time t
+    in
+    let (scope_date, scope_region) = scope in
+    let scope_str = Printf.sprintf "%s/%s/%s/aws4_request"
+        (Date.to_string_iso8601_basic scope_date)
+        (string_of_region scope_region)
+        (string_of_service service)
+    in
+    let hashed_req = digest canonical_request in
+    Printf.sprintf "AWS4-HMAC-SHA256\n%s\n%s\n%s" time_str scope_str hashed_req
+
+  let make_signing_key ?date ~region ~service ~secret_access_key =
+    let mac k v = Nocrypto.Hash.(mac `SHA256
+                                   ~key:k
+                                   (Cstruct.of_string v)) in
+    let date' = match date with
+        None -> Date.today ~zone:Time.Zone.utc
+      | Some d -> d in
+    let date_str = Date.to_string_iso8601_basic date' in
+    let date_key = mac (Cstruct.of_string ("AWS4"^secret_access_key)) date_str in
+    let date_region_key = mac date_key (string_of_region region) in
+    let date_region_service_key = mac date_region_key (string_of_service service) in
+    let signing_key = mac date_region_service_key "aws4_request" in
+    signing_key
+
+  let auth_request ?now ~hashed_payload ~region ~service ~aws_access_key ~aws_secret_key request =
+    (* Important use the same time for everything here *)
+    let time = match now with None -> (Time.now ()) | Some t -> t in
```

Comment:
`Option.value` should simplify this a little (never mind that `Time.now` will always be evaluated)